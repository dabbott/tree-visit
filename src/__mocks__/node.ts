import { IndexPath } from '../indexPath'

export type Node = {
  name: string
  children?: Node[]
  indexPath: IndexPath
}

export function getChildren(node: Node): Node[] {
  return node.children ?? []
}

// Create a new child object for each node to test that we're not relying on object identity
export function getChildrenUnstable(node: Node): Node[] {
  return getChildren(node).map((item) => ({ ...item }))
}

export function getLabel(node: Node): string {
  return node.name
}

export const example: Node = {
  name: 'a',
  children: [
    {
      name: 'b',
      indexPath: [0],
      children: [
        { name: 'b1', indexPath: [0, 0] },
        { name: 'b2', indexPath: [0, 1] },
      ],
    },
    {
      name: 'c',
      indexPath: [1],
      children: [
        { name: 'c1', indexPath: [1, 0] },
        { name: 'c2', indexPath: [1, 1] },
      ],
    },
  ],
  indexPath: [],
}
