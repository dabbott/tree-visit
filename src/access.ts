import { convertChildrenToEntries, getEntriesChild } from './getChild'
import { IndexPath, KeyPath } from './indexPath'
import { BaseEntriesOptions, BaseOptions } from './options'

type AccessChildrenOptions<T> = BaseOptions<T> & {
  indexPath: IndexPath
}

type AccessEntriesOptions<T> = BaseEntriesOptions<T> & {
  keyPath: KeyPath
}

export type AccessOptions<T> =
  | AccessChildrenOptions<T>
  | AccessEntriesOptions<T>

/**
 * Returns a node by its `IndexPath`.
 *
 * The first node is implicitly included in the `IndexPath` (i.e. no need to pass a `0` first in every `IndexPath`).
 */
export function access<T>(node: T, options: AccessChildrenOptions<T>): T
export function access<T>(node: T, options: AccessEntriesOptions<T>): T
export function access<T>(node: T, options: AccessOptions<T>): T {
  const opts =
    'getChildren' in options
      ? { ...convertChildrenToEntries<T>(options), keyPath: options.indexPath }
      : options

  return accessInternal(node, opts)
}

export function accessInternal<T>(
  node: T,
  options: AccessEntriesOptions<T>
): T {
  let path = options.keyPath.slice()

  while (path.length > 0) {
    let key = path.shift()!

    node = getEntriesChild(node, {
      ...options,
      keyPath: path,
      key,
    })
  }

  return node
}

/**
 * Returns an array of each node in an `IndexPath`.
 *
 * The first node is implicitly included in the `IndexPath` (i.e. no need to pass a `0` first in every `IndexPath`).
 */
export function accessPath<T>(node: T, options: AccessChildrenOptions<T>): T[]
export function accessPath<T>(node: T, options: AccessEntriesOptions<T>): T[]
export function accessPath<T>(node: T, options: AccessOptions<T>): T[] {
  const opts =
    'getChildren' in options
      ? { ...convertChildrenToEntries<T>(options), keyPath: options.indexPath }
      : options

  return accessPathInternal(node, opts)
}

export function accessPathInternal<T>(
  node: T,
  options: AccessEntriesOptions<T>
): T[] {
  let result: T[] = [node]
  let path = options.keyPath.slice()

  while (path.length > 0) {
    let key = path.shift()!

    node = getEntriesChild(node, {
      ...options,
      keyPath: path,
      key,
    })

    result.push(node)
  }

  return result
}
