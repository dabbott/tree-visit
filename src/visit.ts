import { convertChildrenToEntries } from './getChild'
import { IndexPath, KeyPath } from './indexPath'
import { BaseChildrenOptions, BaseEntriesOptions } from './options'

export const SKIP = 'skip'
export const STOP = 'stop'

export type EnterReturnValue = void | 'skip' | 'stop'
export type LeaveReturnValue = void | 'stop'
export type VisitOptions<T> = BaseChildrenOptions<T> & {
  onEnter?(node: T, indexPath: IndexPath): EnterReturnValue
  onLeave?(node: T, indexPath: IndexPath): LeaveReturnValue
}
export type VisitEntriesOptions<T> = BaseEntriesOptions<T> & {
  onEnter?(node: T, keyPath: KeyPath): EnterReturnValue
  onLeave?(node: T, keyPath: KeyPath): LeaveReturnValue
}

type NodeWrapper<T> = {
  node: T

  /**
   * The current traversal state of the node.
   *
   * undefined => not visited
   * -1        => skipped
   * n         => nth child
   */
  state?: number

  /**
   * Cached children, so we only call getChildren once per node
   */
  entries?: [PropertyKey, T][]
}

/**
 * Visit each node in the tree, calling an optional `onEnter` and `onLeave` for each.
 *
 * From `onEnter`:
 *
 * - return nothing or `undefined` to continue
 * - return `"skip"` to skip the children of that node and the subsequent `onLeave`
 * - return `"stop"` to end traversal
 *
 * From `onLeave`:
 *
 * - return nothing or `undefined` to continue
 * - return `"stop"` to end traversal
 */
export function visit<T>(node: T, options: VisitOptions<T>): void
export function visit<T>(node: T, options: VisitEntriesOptions<T>): void
export function visit<T>(
  node: T,
  _options: VisitOptions<T> | VisitEntriesOptions<T>
): void {
  const options = visitOptionsInterop(_options)
  const { onEnter, onLeave, getEntries } = options

  let keyPath: KeyPath = []
  let stack: NodeWrapper<T>[] = [{ node }]

  const getKeyPath = options.reuseIndexPath
    ? () => keyPath
    : () => keyPath.slice()

  while (stack.length > 0) {
    let wrapper = stack[stack.length - 1]

    // Visit the wrapped node
    if (wrapper.state === undefined) {
      const enterResult = onEnter?.(wrapper.node, getKeyPath())

      if (enterResult === STOP) return

      wrapper.state = enterResult === SKIP ? -1 : 0
    }

    const entries = wrapper.entries || getEntries(wrapper.node, getKeyPath())

    if (!wrapper.entries) {
      wrapper.entries = entries
    }

    // If the node wasn't skipped
    if (wrapper.state !== -1) {
      // Visit the next child
      if (wrapper.state < entries.length) {
        let currentIndex = wrapper.state

        keyPath.push(entries[currentIndex][0])
        stack.push({ node: entries[currentIndex][1] })

        wrapper.state = currentIndex + 1

        continue
      }

      const leaveResult = onLeave?.(wrapper.node, getKeyPath())

      if (leaveResult === STOP) return
    }

    keyPath.pop()
    stack.pop()
  }
}

function visitOptionsInterop<T>(
  options: VisitOptions<T> | VisitEntriesOptions<T>
): VisitEntriesOptions<T> {
  if ('getEntries' in options) return options

  return {
    ...convertChildrenToEntries<T>(options),
    ...(options.onEnter && { onEnter: options.onEnter }),
    ...(options.onLeave && { onLeave: options.onLeave }),
  }
}
