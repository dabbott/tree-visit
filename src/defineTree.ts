import { access, accessPath } from './access'
import { diagram, DiagramOptions } from './diagram'
import {
  find,
  findAll,
  findAllIndexPaths,
  findIndexPath,
  FindOptions,
  FindOptionsTyped,
} from './find'
import { flat } from './flat'
import { flatMap, FlatMapOptions } from './flatMap'
import { IndexPath } from './indexPath'
import { insert, InsertOptions } from './insert'
import { map, MapOptions } from './map'
import { move, MoveOptions } from './move'
import { BaseOptions, MutationBaseOptions, TraversalContext } from './options'
import { reduce, ReduceOptions } from './reduce'
import { remove, RemoveOptions } from './remove'
import { replace, ReplaceOptions } from './replace'
import { ExtractRequiredKeys, OptionCheck, Prettify } from './types'
import { visit, VisitOptions } from './visit'

type WithoutBase<T> = Omit<T, keyof BaseOptions<T>>

type MutationOptions<T> = WithoutBase<MutationBaseOptions<T>>

type DiagramOptionsWB<T> = WithoutBase<DiagramOptions<T>>
type DiagramRequiredOptions<T> = Pick<
  DiagramOptionsWB<T>,
  ExtractRequiredKeys<DiagramOptionsWB<T>>
>
type DiagramOptionalOptions<T> = Omit<
  DiagramOptionsWB<T>,
  ExtractRequiredKeys<DiagramOptionsWB<T>>
>
type FindOptionsWB<T> = WithoutBase<FindOptions<T>>
type VisitOptionsWB<T> = WithoutBase<VisitOptions<T>>
type InsertOptionsWB<T> = WithoutBase<InsertOptions<T>>
type RemoveOptionsWB<T> = WithoutBase<RemoveOptions<T>>
type MoveOptionsWB<T> = WithoutBase<MoveOptions<T>>
type ReplaceOptionsWB<T> = WithoutBase<ReplaceOptions<T>>

type ApplyableOptions<T> = DiagramRequiredOptions<T> & MutationOptions<T>

interface Overloads<T> {
  /**
   * Find a node matching a predicate function.
   */
  find(node: T, predicate: FindOptions<T>['predicate']): T | undefined

  find(node: T, options: FindOptionsWB<T>): T | undefined

  find<S extends T>(
    node: T,
    predicate: FindOptionsTyped<T, S>['predicate']
  ): S | undefined

  find<S extends T>(
    node: T,
    options: WithoutBase<FindOptionsTyped<T, S>>
  ): S | undefined

  /**
   * Find all nodes matching a predicate function.
   */
  findAll(node: T, predicate: FindOptions<T>['predicate']): T[]

  findAll(node: T, options: FindOptionsWB<T>): T[]

  findAll<S extends T>(
    node: T,
    predicate: FindOptionsTyped<T, S>['predicate']
  ): S[]

  findAll<S extends T>(
    node: T,
    options: WithoutBase<FindOptionsTyped<T, S>>
  ): S[]

  /**
   * Find the `IndexPath` of a node matching a predicate function.
   */
  findIndexPath(
    node: T,
    predicate: FindOptions<T>['predicate']
  ): IndexPath | undefined

  findIndexPath(node: T, options: FindOptionsWB<T>): IndexPath | undefined

  /**
   * Find the `IndexPath` of all nodes matching a predicate function.
   */
  findAllIndexPaths(
    node: T,
    predicate: FindOptions<T>['predicate']
  ): IndexPath[]

  findAllIndexPaths(node: T, options: FindOptionsWB<T>): IndexPath[]

  /**
   * Visit each node using preorder DFS traversal.
   */
  visit(node: T, onEnter: NonNullable<VisitOptions<T>>['onEnter']): void

  /**
   * Visit each node using DFS traversal.
   */
  visit(node: T, options: VisitOptionsWB<T>): void
}

class Tree<T, AppliedOptions extends Partial<ApplyableOptions<T>>> {
  constructor(
    getChildrenOrBaseOptions: BaseOptions<T> | BaseOptions<T>['getChildren'],
    public appliedOptions: AppliedOptions
  ) {
    this.baseOptions =
      typeof getChildrenOrBaseOptions === 'function'
        ? { getChildren: getChildrenOrBaseOptions }
        : getChildrenOrBaseOptions

    this._getChildren = this.baseOptions.getChildren
  }

  _getChildren: BaseOptions<T>['getChildren']

  /**
   * Returns the node's children.
   *
   * This is the same as the `getChildren` option passed to `defineTree`, included here for convenience.
   */
  getChildren = (
    node: T,
    indexPath: IndexPath,
    context?: TraversalContext<T>
  ) => {
    return this._getChildren(node, indexPath, context)
  }

  baseOptions: BaseOptions<T>

  mergeOptions = <T extends Record<string, any>>(options: T) => ({
    ...this.baseOptions,
    ...this.appliedOptions,
    ...options,
  })

  withOptions = <NewOptions extends Partial<ApplyableOptions<T>>>(
    newOptions: NewOptions
  ) => new Tree(this.baseOptions, { ...this.appliedOptions, ...newOptions })

  /**
   * Returns a node by its `IndexPath`.
   *
   * The first node is implicitly included in the `IndexPath` (i.e. no need to pass a `0` first in every `IndexPath`).
   */
  access = (node: T, indexPath: IndexPath) =>
    access(node, indexPath, this.mergeOptions({}))

  /**
   * Returns an array of each node in an `IndexPath`.
   *
   * The first node is implicitly included in the `IndexPath` (i.e. no need to pass a `0` first in every `IndexPath`).
   */
  accessPath = (node: T, indexPath: IndexPath) =>
    accessPath(node, indexPath, this.mergeOptions({}))

  /**
   * Generate a diagram of the tree, as a string.
   */
  diagram = (
    node: T,
    options:
      | DiagramRequiredOptions<T>['getLabel']
      | Prettify<
          AppliedOptions extends DiagramRequiredOptions<T>
            ? DiagramOptionalOptions<T> | void
            : OptionCheck<
                AppliedOptions,
                'getLabel',
                DiagramRequiredOptions<T>
              > &
                DiagramOptionalOptions<T>
        >
  ) =>
    typeof options === 'function'
      ? diagram(node, this.mergeOptions({ getLabel: options }))
      : diagram(node, this.mergeOptions(options))

  find: Overloads<T>['find'] = (
    node: T,
    predicateOrOptions: FindOptions<T>['predicate'] | FindOptionsWB<T>
  ) =>
    typeof predicateOrOptions === 'function'
      ? find(node, this.mergeOptions({ predicate: predicateOrOptions }))
      : find(node, this.mergeOptions({ ...predicateOrOptions }))

  findAll: Overloads<T>['findAll'] = (
    node: T,
    predicateOrOptions: FindOptions<T>['predicate'] | FindOptionsWB<T>
  ) =>
    typeof predicateOrOptions === 'function'
      ? findAll(node, this.mergeOptions({ predicate: predicateOrOptions }))
      : findAll(node, this.mergeOptions({ ...predicateOrOptions }))

  findIndexPath: Overloads<T>['findIndexPath'] = (
    node: T,
    predicateOrOptions: FindOptions<T>['predicate'] | FindOptionsWB<T>
  ) =>
    typeof predicateOrOptions === 'function'
      ? findIndexPath(
          node,
          this.mergeOptions({ predicate: predicateOrOptions })
        )
      : findIndexPath(node, this.mergeOptions({ ...predicateOrOptions }))

  findAllIndexPaths: Overloads<T>['findAllIndexPaths'] = (
    node: T,
    predicateOrOptions: FindOptions<T>['predicate'] | FindOptionsWB<T>
  ) =>
    typeof predicateOrOptions === 'function'
      ? findAllIndexPaths(
          node,
          this.mergeOptions({ predicate: predicateOrOptions })
        )
      : findAllIndexPaths(node, this.mergeOptions({ ...predicateOrOptions }))

  /**
   * Returns an array containing the root node and all of its descendants.
   *
   * This is analogous to `Array.prototype.flat` for flattening arrays.
   */
  flat = (node: T) => flat(node, this.mergeOptions({}))

  /**
   * Map each node into an array of values, which are then flattened into a single array.
   *
   * This is analogous to `Array.prototype.flatMap` for arrays.
   */
  flatMap = <R>(node: T, transform: FlatMapOptions<T, R>['transform']) =>
    flatMap(node, this.mergeOptions({ transform }))

  reduce = <R>(
    node: T,
    nextResult: ReduceOptions<T, R>['nextResult'],
    initialResult: R
  ): R => reduce(node, this.mergeOptions({ nextResult, initialResult }))

  map = <R>(node: T, transform: MapOptions<T, R>['transform']): R =>
    map(node, this.mergeOptions({ transform }))

  visit: Overloads<T>['visit'] = (
    node: T,
    onEnterOrOptions:
      | NonNullable<VisitOptions<T>>['onEnter']
      | VisitOptionsWB<T>
  ) =>
    typeof onEnterOrOptions === 'function'
      ? visit(node, this.mergeOptions({ onEnter: onEnterOrOptions }))
      : visit(node, this.mergeOptions({ ...onEnterOrOptions }))

  // --- Mutations ---

  /**
   * Insert nodes at a given `IndexPath`.
   */
  insert = (
    node: T,
    options: Prettify<
      OptionCheck<AppliedOptions, 'create', InsertOptionsWB<T>> &
        Omit<InsertOptionsWB<T>, 'create'>
    >
  ) => insert(node, this.mergeOptions(options))

  /**
   * Remove nodes at the given `IndexPath`s.
   */
  remove = (
    node: T,
    options: Prettify<
      OptionCheck<AppliedOptions, 'create', RemoveOptionsWB<T>> &
        Omit<RemoveOptionsWB<T>, 'create'>
    >
  ) => remove(node, this.mergeOptions(options))

  /**
   * Move nodes from one `IndexPath` to another.
   */
  move = (
    node: T,
    options: Prettify<
      OptionCheck<AppliedOptions, 'create', MoveOptionsWB<T>> &
        Omit<MoveOptionsWB<T>, 'create'>
    >
  ) => move(node, this.mergeOptions(options))

  /**
   * Replace the node at the given `IndexPath` with another
   */
  replace = (
    node: T,
    options: Prettify<
      OptionCheck<AppliedOptions, 'create', ReplaceOptionsWB<T>> &
        Omit<ReplaceOptionsWB<T>, 'create'>
    >
  ) => replace(node, this.mergeOptions(options))
}

export function defineTree<T>(
  getChildren: BaseOptions<T> | BaseOptions<T>['getChildren']
) {
  return new Tree(getChildren, {})
}
