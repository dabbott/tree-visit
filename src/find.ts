import { IndexPath } from './indexPath'
import {
  BaseChildrenOptions,
  BaseEntriesOptions,
  convertChildrenToEntries,
} from './options'
import { STOP, visit } from './visit'

export type FindChildrenOptions<T> = BaseChildrenOptions<T> & {
  /**
   * Return `true` to include this node in the results.
   */
  predicate: (node: T, indexPath: IndexPath) => boolean
}

export type FindChildrenOptionsTyped<
  T,
  S extends T
> = BaseChildrenOptions<T> & {
  /**
   * Return `true` to include this node in the results.
   */
  predicate: (node: T, indexPath: IndexPath) => node is S
}

export type FindEntriesOptions<T, PK extends PropertyKey> = BaseEntriesOptions<
  T,
  PK
> & {
  /**
   * Return `true` to include this node in the results.
   */
  predicate: (node: T, keyPath: PK[]) => boolean
}

export type FindEntriesOptionsTyped<
  T,
  PK extends PropertyKey,
  S extends T
> = BaseEntriesOptions<T, PK> & {
  /**
   * Return `true` to include this node in the results.
   */
  predicate: (node: T, keyPath: PK[]) => node is S
}

/**
 * Find a node matching a predicate function.
 */
export function find<T>(node: T, options: FindChildrenOptions<T>): T | undefined
export function find<T, S extends T>(
  node: T,
  options: FindChildrenOptionsTyped<T, S>
): S | undefined
export function find<T, PK extends PropertyKey>(
  node: T,
  options: FindEntriesOptions<T, PK>
): T | undefined
export function find<T, PK extends PropertyKey, S extends T>(
  node: T,
  options: FindEntriesOptionsTyped<T, PK, S>
): S | undefined
export function find<T, PK extends PropertyKey>(
  node: T,
  _options: FindChildrenOptions<T> | FindEntriesOptions<T, PK>
): T | undefined {
  const options = findOptionsInterop<T, PK>(_options)

  let found: T | undefined

  visit(node, {
    ...options,
    onEnter: (child, indexPath) => {
      if (options.predicate(child, indexPath)) {
        found = child
        return STOP
      }
    },
  })

  return found
}

/**
 * Find all nodes matching a predicate function.
 */
export function findAll<T>(node: T, options: FindChildrenOptions<T>): T[]
export function findAll<T, S extends T>(
  node: T,
  options: FindChildrenOptionsTyped<T, S>
): S[]
export function findAll<T>(node: T, options: FindChildrenOptions<T>): T[] {
  let found: T[] = []

  visit(node, {
    onEnter: (child, indexPath) => {
      if (options.predicate(child, indexPath)) {
        found.push(child)
      }
    },
    getChildren: options.getChildren,
  })

  return found
}

/**
 * Find the `IndexPath` of a node matching a predicate function.
 */
export function findIndexPath<T>(
  node: T,
  options: FindChildrenOptions<T>
): IndexPath | undefined {
  let found: IndexPath | undefined

  visit(node, {
    onEnter: (child, indexPath) => {
      if (options.predicate(child, indexPath)) {
        // Copy the indexPath, since indexPath may be mutated
        found = [...indexPath]
        return STOP
      }
    },
    getChildren: options.getChildren,
  })

  return found
}

/**
 * Find the `IndexPath` of all nodes matching a predicate function.
 */
export function findAllIndexPaths<T>(
  node: T,
  options: FindChildrenOptions<T>
): IndexPath[] {
  let found: IndexPath[] = []

  visit(node, {
    onEnter: (child, indexPath) => {
      if (options.predicate(child, indexPath)) {
        // Copy the indexPath, since indexPath may be mutated
        found.push([...indexPath])
      }
    },
    getChildren: options.getChildren,
  })

  return found
}

function findOptionsInterop<T, PK extends PropertyKey>(
  options: FindChildrenOptions<T> | FindEntriesOptions<T, PK>
): FindEntriesOptions<T, PK> {
  if ('getEntries' in options) return options

  return {
    ...convertChildrenToEntries<T, PK>(options),
    ...(options.predicate && {
      predicate: options.predicate as FindEntriesOptions<T, PK>['predicate'],
    }),
  }
}
