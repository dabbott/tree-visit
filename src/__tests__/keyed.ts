import { Node, example } from '../__mocks__/node'
import { access, accessPath } from '../access'
import { defineTree } from '../defineTree'
import { find } from '../find'
import { visit } from '../visit'

const getEntries = (node: Node) =>
  (node.children || []).map((n): [string, Node] => [n.name, n])

describe('keyed access', () => {
  it('accesses by key', () => {
    expect(
      access(example, {
        getEntries,
        path: ['b', 'b2'],
      }).name
    ).toEqual('b2')
  })

  it('accesses path by key', () => {
    expect(
      accessPath(example, {
        getEntries,
        path: ['b', 'b2'],
      }).map((n) => n.name)
    ).toEqual(['a', 'b', 'b2'])
  })

  it('visits by key', () => {
    const result: string[] = []

    visit(example, {
      getEntries,
      onEnter(node) {
        result.push(node.name)
      },
    })

    expect(result).toEqual(['a', 'b', 'b1', 'b2', 'c', 'c1', 'c2'])
  })

  it('finds by key', () => {
    expect(
      find(example, {
        getEntries,
        predicate: (node) => node.name === 'b2',
      })?.name
    ).toEqual('b2')
  })
})

describe('keyed partially applied', () => {
  const { diagram, access, find } = defineTree({ getEntries }).withOptions({
    getLabel: (node, keyPath) =>
      node.name + ': ' + [example.name, ...keyPath].join('/'),
  })

  it('renders diagram', () => {
    expect(diagram(example)).toMatchSnapshot()
  })

  it('accesses', () => {
    expect(access(example, ['b', 'b2']).name).toEqual('b2')
  })

  it('finds', () => {
    expect(find(example, (node) => node.name === 'b2')?.name).toEqual('b2')

    expect(
      find(example, {
        predicate: (node) => node.name === 'b2',
      })?.name
    ).toEqual('b2')
  })
})
