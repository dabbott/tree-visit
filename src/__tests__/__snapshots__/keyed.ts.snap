// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keyed partially applied renders diagram 1`] = `
"a: a
├── b: a/b
│   ├── b1: a/b/b1
│   └── b2: a/b/b2
└── c: a/c
    ├── c1: a/c/c1
    └── c2: a/c/c2"
`;
